### PR TITLE
[WIP] UTY-709 Implement base for Readers

### PR DESCRIPTION
**Pull request by firtina-improbable**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
Very WIP, very subject to change, pls look at it only when you are added into the PR

#### Tests
How did you test these changes prior to submitting this pull request?
What automated tests are included in this PR?

#### Documentation
How is this documented (for example: release note, upgrade guide, feature page, in-code documentation)?

#### Primary reviewers
If your change will take a long time to review, you can name at most two primary reviewers who are ultimately responsible for reviewing this request. @ mention them.

### CODE REVIEW

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/IReaderInternal.cs [hunk 1]
```diff
@@ -1,4 +1,7 @@
-﻿using Improbable.Worker;
+﻿using System.Runtime.CompilerServices;
+using Improbable.Worker;
+
+[assembly: InternalsVisibleTo("Improbable.Gdk.Core.EditmodeTests")]
```
Comment (JonasImprobable) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/IReaderInternal.cs]: What is this for? Is this here because `Improbable.Gdk.Core.MonoBehaviours` is otherwise not visible in the editmode tests?

Comment (firtina-improbable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/IReaderInternal.cs]: This is because tests such as `AuthorityChanged_gets_triggered_when_the_reader_receives_an_authority_change` need to use the IReaderInternal interface, but we do not want non-core modules to see it

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/ReaderBase.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+namespace Improbable.Gdk.Core.MonoBehaviours
+{
+    public abstract class ReaderBase<TSpatialComponentData, TComponentUpdate>
+        : IReader<TSpatialComponentData>,
+            IReaderInternal
+        where TSpatialComponentData : ISpatialComponentData
+        where TComponentUpdate : ISpatialComponentUpdate<TSpatialComponentData>
+    {
+        protected readonly Unity.Entities.Entity Entity;
```
Comment (JonasImprobable) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/ReaderBase.cs]: Is there a way to get rid of the `Unity.Entities` prefix? Maybe something like `using Entity = Unity.Entities.Entity;`?

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/ReaderBase.cs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+                {
+                    return Authority.Authoritative;
+                }
+
+                if (EntityManager.HasComponent<AuthorityLossImminent<TSpatialComponentData>>(Entity))
+                {
+                    return Authority.AuthorityLossImminent;
+                }
+
+                // TODO reviewers: should this throw an error instead?
```
Comment (JonasImprobable) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/ReaderBase.cs]: We should throw a "this should never happen" sort of error for the case where no authority tag is on an entity.

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/ReaderBase.cs [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        {
+            add => componentUpdateDelegates.Add(value);
+            remove => componentUpdateDelegates.Remove(value);
+        }
+
+        void IReaderInternal.OnAuthorityChange(Authority authority)
+        {
+            foreach (var authorityChangedDelegate in authorityChangedDelegates)
+            {
+                // TODO catch errors here?
```
Comment (JonasImprobable) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/ReaderBase.cs]: Yes, let's wrap any user code in try catch blocks.

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/IReaderInternal.cs [hunk 5]
```diff
@@ -1,6 +1,9 @@
-﻿using Improbable.Worker;
+﻿using System.Runtime.CompilerServices;
+using Improbable.Worker;
 
-namespace Improbable.Gdk.Core.MonoBehaviours
+[assembly: InternalsVisibleTo("Improbable.Gdk.Core.EditmodeTests")]
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/IReaderInternal.cs]: Move this into its own `AssemblyInfo.cs`, since it's global to the entire assembly.

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/BlittableReaderBase.cs [hunk 6]
```diff
@@ -0,0 +1,16 @@
+﻿using Unity.Entities;
+
+namespace Improbable.Gdk.Core
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/BlittableReaderBase.cs]: Why isn't this in a sub-namespace, or better yet, not in Core at all?

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/ReaderBase.cs [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+                {
+                    return Authority.AuthorityLossImminent;
+                }
+
+                if (EntityManager.HasComponent<NotAuthoritative<TSpatialComponentData>>(Entity))
+                {
+                    return Authority.NotAuthoritative;
+                }
+
+                throw new Exception("No authority component found for this entity.");
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/ReaderBase.cs]: What happened? How can the user fix it?

File: workers/unity/Packages/com.improbable.gdk.core/Tests/Editmode/GameObjectRepresentation/Readers/ReaderDataTests.cs [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace Improbable.Gdk.Core.EditmodeTests.Readers
+{
+    [TestFixture]
+    public class ReaderDataTests : ReaderTestsBase
+    {
+        [Test]
+        public void Data_returns_component_instance()
+        {
+            EntityManager.SetComponentData(Entity, new ReaderTestComponent { FloatValue = 5 });
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.core/Tests/Editmode/GameObjectRepresentation/Readers/ReaderDataTests.cs]: nit: `5.0f`

File: workers/unity/Packages/com.improbable.gdk.core/Tests/Editmode/GameObjectRepresentation/Readers/ReaderTests.cs [hunk 9]
```diff
@@ -0,0 +1,111 @@
+﻿using System;
+using System.Collections.Generic;
+using NUnit.Framework;
+using Unity.Entities;
+using UnityEngine;
+using Entity = Unity.Entities.Entity;
+
+namespace Improbable.Gdk.Core.EditmodeTests.Readers
+{
+    public abstract class ReaderTestsBase
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.core/Tests/Editmode/GameObjectRepresentation/Readers/ReaderTests.cs]: Couldn't this be code-genned?

File: workers/unity/Packages/com.improbable.gdk.core/Tests/Editmode/GameObjectRepresentation/Readers/ReaderUpdateTests.cs [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+        private void QueueUpdatesToEntity(params ReaderTestComponent.Update[] updatesToSend)
+        {
+            if (!EntityManager.HasComponent<ComponentsUpdated<ReaderTestComponent.Update>>(Entity))
+            {
+                EntityManager.AddComponent(Entity, typeof(ComponentsUpdated<ReaderTestComponent.Update>));
+
+                var componentsUpdated = new ComponentsUpdated<ReaderTestComponent.Update>();
+                EntityManager.SetComponentObject(Entity, componentsUpdated);
+            }
+
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.core/Tests/Editmode/GameObjectRepresentation/Readers/ReaderUpdateTests.cs]: nit: extra line

File: workers/unity/Packages/com.improbable.gdk.core/Tests/Editmode/GameObjectRepresentation/Readers/ReaderUpdateTests.cs [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+        {
+            bool floatValueUpdated = false;
+            bool intValueUpdated = false;
+
+            Reader.FloatValueUpdated += newValue => { floatValueUpdated = true; };
+            Reader.IntValueUpdated += newValue => { intValueUpdated = true; };
+
+            QueueUpdatesToEntity(new ReaderTestComponent.Update
+            {
+                FloatValue = new Option<float>(10),
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.core/Tests/Editmode/GameObjectRepresentation/Readers/ReaderUpdateTests.cs]: Nit: `10.0f`

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/BlittableReaderBase.cs (hunk 6 above)
Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/BlittableReaderBase.cs]: Yeah please put this in `Improbable.Gdk.Core.Monobehaviours` or similar

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/BlittableReaderBase.cs [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+{
+    public abstract class BlittableReaderBase<TSpatialComponentData, TComponentUpdate>
+        : ReaderBase<TSpatialComponentData, TComponentUpdate>
+        where TSpatialComponentData : struct, ISpatialComponentData, IComponentData
+        where TComponentUpdate : ISpatialComponentUpdate<TSpatialComponentData>
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/BlittableReaderBase.cs]: Again, is this second generic constraint required? It doesn't actually provide any extra constraints?

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/BlittableReaderBase.cs [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+    {
+        protected BlittableReaderBase(Entity entity, EntityManager entityManager) : base(entity, entityManager)
+        {
+        }
+
+        public override TSpatialComponentData Data => EntityManager.GetComponentData<TSpatialComponentData>(Entity);
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/BlittableReaderBase.cs]: This can throw exceptions, we should handle that case properly

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/IReader.cs [hunk 14]
```diff
@@ -1,7 +1,6 @@
 ﻿using Improbable.Worker;
-using Unity.Entities;
 
-namespace Improbable.Gdk.Core.MonoBehaviours
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/IReader.cs]: Why are we moving everything into just `Core`

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/NonBlittableReaderBase.cs [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+        : ReaderBase<TSpatialComponentData, TComponentUpdate>
+        where TSpatialComponentData : Component, ISpatialComponentData
+        where TComponentUpdate : ISpatialComponentUpdate<TSpatialComponentData>
+    {
+        protected NonBlittableReaderBase(Entity entity, EntityManager entityManager) : base(entity, entityManager)
+        {
+        }
+
+        public override TSpatialComponentData Data => EntityManager.GetComponentObject<TSpatialComponentData>(Entity);
+    }
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/NonBlittableReaderBase.cs]: Same comments apply on this file as the Blittable one before

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/ReaderBase.cs (hunk 7 above)
Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/ReaderBase.cs]: They can't - its a failing in the core somewhere

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/ReaderBase.cs [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+            foreach (var authorityChangedDelegate in authorityChangedDelegates)
+            {
+                try
+                {
+                    authorityChangedDelegate(authority);
+                }
+                catch (Exception e)
+                {
+                    // Log the exception but do not rethrow it, as other delegates should still get called
+                    Debug.LogException(e);
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/ReaderBase.cs]: Should we not use the LogDispatcher?

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/ReaderBase.cs [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+        private readonly List<GameObjectDelegates.ComponentUpdated<TSpatialComponentData>> componentUpdateDelegates
+            = new List<GameObjectDelegates.ComponentUpdated<TSpatialComponentData>>();
+
+        public event GameObjectDelegates.ComponentUpdated<TSpatialComponentData> ComponentUpdated
+        {
+            add => componentUpdateDelegates.Add(value);
+            remove => componentUpdateDelegates.Remove(value);
+        }
+
+        void IReaderInternal.OnAuthorityChange(Authority authority)
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/ReaderBase.cs]: Needs an access modifier

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/ReaderBase.cs [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+                }
+                catch (Exception e)
+                {
+                    // Log the exception but do not rethrow it, as other delegates should still get called
+                    Debug.LogException(e);
+                }
+            }
+        }
+
+        void IReaderInternal.OnComponentUpdate()
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/ReaderBase.cs]: Needs access modifier

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/ReaderBase.cs [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+                foreach (var componentUpdateDelegate in componentUpdateDelegates)
+                {
+                    try
+                    {
+                        componentUpdateDelegate(update);
+                    }
+                    catch (Exception e)
+                    {
+                        // Log the exception but do not rethrow it, as other delegates should still get called
+                        Debug.LogException(e);
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/ReaderBase.cs]: LogDispatcher?

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/ReaderBase.cs [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+
+        /// <summary>
+        ///     Reader implementations will override this if their components have fields.
+        /// </summary>
+        /// <param name="update"></param>
+        protected virtual void HandleFieldUpdates(TComponentUpdate update)
+        {
+        }
+
+        void IReaderInternal.OnEvent(int eventIndex)
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/ReaderBase.cs]: Needs access modifier

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/ReaderBase.cs [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+        {
+            // TODO
+        }
+
+        void IReaderInternal.OnCommandRequest(int commandIndex)
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/ReaderBase.cs]: Needs access modifier

Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/ReaderBase.cs]: FYI - you won't have the commandIndex until `13.1` as the JSON AST doesn't have it until `13.1`

File: workers/unity/Packages/com.improbable.gdk.core/Tests/Editmode/GameObjectRepresentation/Readers/NonBlittableReaderTests.cs [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+                var entityManager = world.GetOrCreateManager<EntityManager>();
+
+                var entity = entityManager.CreateEntity(typeof(NonBlittableTestComponent));
+
+                var reader = new NonBlittableTestComponent.Reader(entity, entityManager);
+
+                entityManager.SetComponentObject(entity, new NonBlittableTestComponent());
+
+                var data = entityManager.GetComponentObject<NonBlittableTestComponent>(entity);
+
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.core/Tests/Editmode/GameObjectRepresentation/Readers/NonBlittableReaderTests.cs]: Nit: A loot of unnecessary whitespace imo

File: workers/unity/Packages/com.improbable.gdk.core/Tests/Editmode/GameObjectRepresentation/Readers/ReaderAuthorityTests.cs [hunk 23]
```diff
@@ -0,0 +1,88 @@
...
+        public void Authority_throws_if_the_entity_has_no_authority_components()
+        {
+            // entity is fresh so no components on it
+            Assert.Throws<Exception>(() => { Debug.Log(Reader.Authority); });
+        }
+
+        [Test]
+        public void Authority_returns_NotAuthoritative_when_no_authority_components_are_present()
+        {
+            EntityManager.AddComponent(Entity, typeof(NotAuthoritative<ReaderTestComponent>));
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.core/Tests/Editmode/GameObjectRepresentation/Readers/ReaderAuthorityTests.cs]: Should we not remove this after the assert to clean the entity?

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/ReaderBase.cs [hunk 24]
```diff
@@ -0,0 +1,123 @@
...
+        public Authority Authority
+        {
+            get
+            {
+                if (EntityManager.HasComponent<Authoritative<TSpatialComponentData>>(Entity))
+                {
+                    return Authority.Authoritative;
+                }
+
+                if (EntityManager.HasComponent<AuthorityLossImminent<TSpatialComponentData>>(Entity))
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/ReaderBase.cs]: If you are `AuthorityLossImminent` you will have both the `Authoritative` and `AuthorityLossImminent` tags - as you are strictly still Authoritative - this allows users to not have to duplicate injections in ECS systems.

Should reorder these ifs to reflect that

File: workers/unity/Packages/com.improbable.gdk.core/Tests/Editmode/GameObjectRepresentation/Readers/ReaderAuthorityTests.cs (hunk 23 above)
Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.core/Tests/Editmode/GameObjectRepresentation/Readers/ReaderAuthorityTests.cs]: Similarly for the other tests too

File: workers/unity/Packages/com.improbable.gdk.core/Tests/Editmode/GameObjectRepresentation/Readers/ReaderAuthorityTests.cs [hunk 25]
```diff
@@ -0,0 +1,88 @@
...
+
+            Assert.AreEqual(true, authorityChangedCalled);
+        }
+
+        [Test]
+        public void AuthorityChanged_calls_non_failure_callbacks_even_if_some_callbacks_fail()
+        {
+            var secondAuthorityChangeCalled = false;
+
+            Reader.AuthorityChanged += authority => throw new Exception("backwards time travel");
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.core/Tests/Editmode/GameObjectRepresentation/Readers/ReaderAuthorityTests.cs]: ⌛️

File: workers/unity/Packages/com.improbable.gdk.core/Tests/Editmode/GameObjectRepresentation/Readers/ReaderUpdateTests.cs [hunk 26]
```diff
@@ -0,0 +1,153 @@
...
+            };
+
+            Assert.AreEqual(false, componentUpdated, "Adding an event callback should not fire it immediately");
+
+            QueueUpdatesToEntity(updateToSend);
+
+            var internalReader = (IReaderInternal) Reader;
+
+            internalReader.OnComponentUpdate();
+
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.core/Tests/Editmode/GameObjectRepresentation/Readers/ReaderUpdateTests.cs]: Nit: Lots of whitespace throughout this file

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/IReader.cs (hunk 14 above)
Comment (MPeti01) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/IReader.cs]: Plan is to move everything out in a separate PR before merging into master as far as I understand (right now existing files are in just Core)

**Comment by firtina-improbable** [comment]
Splitting into multiple PRs :hydra:

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/IReader.cs (hunk 14 above)
Comment (firtina-improbable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/IReader.cs]: https://github.com/spatialos/UnityGDK/pull/141

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/ReaderBase.cs (hunk 7 above)
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/ReaderBase.cs]: So far, we've been creating bespoke exception types for each exception. Please make an AuthorityTagNotFoundException sort of custom exception class.

**Comment by firtina-improbable** [comment]
Multiple PRs, multiple profit
